wsgi/prz-recipe-options.rst

```diff
@@ -23,30 +18,131 @@ The default is to listen on 0.0.0.0:8080.
+    The number of workers is configured in ``wsgi.ini`` (``waitress``) or ``zope.conf`` (``ZServer``).
+    Worker threads are spawned **once** when the Plone instance is starting up.
+    In Zope and Plone each of these workers will also open a ZODB connection.
+    You can check this in the ZMI Control Panel, "Database Management".
+
+    Other WSGI servers like ``bjoern``, ``gunicorn`` and ``uWSGI`` implement different or additional types of workers and they also differ in the way they are clearing the request queue.
+    We will try to cover at least some of the details in later chapters.
 
 wsgi
 ----
 
+This parameters has two forms.
+It can be either ``on`` or ``off`` signaling whether to use the default WSGI server ``waitress`` or not use a WSGI server but ``ZServer`` instead.
+The latter is only an option for Python 2 since ``ZServer`` is not (yet?) available for Python 3.
+
+Alternatively you can specify your own PasteDeploy ``ini`` file with this option.
+No ``wsgi.ini`` configuration will be provided in this case.
+Most likely ``wsgi-ini-template`` will be easier to use however, see below.
+
+wsgi-ini-template
+-----------------
+
+The path to a WSGI configuration file template for a PasteDeploy compatible WSGI server.
+We will use this form later in the training.
+
 debug-exceptions
 ----------------
 
 Use WSGI middleware debugging facilities.
+When set to ``on`` this option will disable exception views and thus propagate errors to the WSGI stack.
+This allows you to use specific debugging WSGI middleware like the `werkzeig debugger <https://werkzeug.palletsprojects.com/en/0.15.x/debug/>`_.
+We'll cover this in the Debugging chapter.
 
-http-fast-listen
-----------------
+access-log, z2-log
+------------------
+
+``access-log`` has been introduced as an alias for the ``z2-log`` option to specify the location of the access log file.
+
+access-log-level, z2-log-level
+------------------------------
+
+``access-log-level`` is a new alias for ``z2-log-level`` to specify the log level of the access log file.
+For WSGI the default is ``INFO`` (``WARN`` for ZServer).
+
+Advanced logging for WSGI
+-------------------------
+
+``access-log-handler``, ``access-log-args``, ``access-log-kwargs``, ``event-log-handler``, ``event-log-args``, ``event-log-kwargs`` are WSGI only options that you can use to configure arbitrary Python logging handlers.
+Numerous logging handlers are defined in the `Python standard library <https://docs.python.org/3/library/logging.handlers.html>`_.
+
+Exercise 1
+++++++++++
+
+Configure Plone access logging to a TCP Server using ``logging.handlers.SocketHandler`` from the Python standard library.
+Use a local address.
+
+
+..  admonition:: Solution
+    :class: toggle
+
+    Add the following two lines to your buildout configuration (we use ``options.cfg``) from the ``wsgitraining_buildout`` as a starting point:
+
+    .. code-block:: ini
+        :emphasize-lines: 9-10
+
+        ...
```

Comment:
`...` is not legal ini syntax. It must be commented so that syntax highlighting works. `#...`

Reply:
not sure about this. Syntax highlighting seems to work fine without the '#', not only on my box but also in the online version of the training, see e.g. here https://training.plone.org/5/wsgi/bjoern.html#exercise-1.

If this is really the case, I'd like to go without the '#' because I think it can be confusing to the reader.

Reply:
If you put a literal `...` in an `.ini`, what happens?

The highlighting might appear plain or it might try to highlight `...` as a literal instead of the intent to indicate that something has been omitted. It might be better to just remove it altogether.

Reply:
Indicating that something has been omitted imo implies that it is unimportant, so it shouldn't ever be necessary to highlight it (certainly I don't want to highlight the ellipsis here). And doesn't the fact that the ellipsis isn't valid `ini` syntax make it even clearer to the reader that the following code is only an extraction of the complete file?

Reply:
When I taught Pyramid tutorials where `...` was in the code, can you guess what the students copy-pasted from the code samples? Can you guess what I removed from the tutorial code samples? I've never had that problem again.